x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineDisjunctions.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    protected static CIDRMatch createCIDRMatch(Expression k, List<Expression> v) {
+        return new CIDRMatch(k.source(), k, v);
+    }
+
+    @Override
+    public Expression rule(Or or) {
+        Expression e = or;
+        // look only at equals, In and CIDRMatch
+        List<Expression> exps = splitOr(e);
+
+        Map<Expression, Set<Expression>> ins = new LinkedHashMap<>();
+        Map<Expression, Set<Expression>> cidrs = new LinkedHashMap<>();
+        Map<Expression, Set<Expression>> ips = new LinkedHashMap<>();
+        ZoneId zoneId = null;
+        List<Expression> ors = new LinkedList<>();
+        boolean changed = false;
+        for (Expression exp : exps) {
+            if (exp instanceof Equals eq) {
+                // consider only equals against foldables
+                if (eq.right().foldable()) {
+                    ins.computeIfAbsent(eq.left(), k -> new LinkedHashSet<>()).add(eq.right());
+                    if (eq.left().dataType() == DataType.IP) {
+                        Object value = eq.right().fold();
+                        if (value instanceof BytesRef bytesRef) {
+                            value = ipToString(bytesRef);
+                        }
+                        ips.computeIfAbsent(eq.left(), k -> new LinkedHashSet<>()).add(new Literal(Source.EMPTY, value, DataType.IP));
+                    }
+                } else {
+                    ors.add(exp);
+                }
+                if (zoneId == null) {
+                    zoneId = eq.zoneId();
+                }
+            } else if (exp instanceof In in) {
+                ins.computeIfAbsent(in.value(), k -> new LinkedHashSet<>()).addAll(in.list());
+                if (in.value().dataType() == DataType.IP) {
+                    List<Expression> values = new ArrayList<>(in.list().size());
+                    for (Expression i : in.list()) {
+                        Object value = i.fold();
+                        if (value instanceof BytesRef bytesRef) {
+                            value = ipToString(bytesRef);
+                        }
+                        values.add(new Literal(Source.EMPTY, value, DataType.IP));
+                    }
+                    ips.computeIfAbsent(in.value(), k -> new LinkedHashSet<>()).addAll(values);
+                }
+            } else if (exp instanceof CIDRMatch cm) {
+                cidrs.computeIfAbsent(cm.ipField(), k -> new LinkedHashSet<>()).addAll(cm.matches());
+            } else {
+                ors.add(exp);
+            }
+        }
+
+        if (cidrs.isEmpty() == false) {
+            for (Expression f : ips.keySet()) {
+                cidrs.computeIfAbsent(f, k -> new LinkedHashSet<>()).addAll(ips.get(f));
+                ins.remove(f);
+            }
+        }
+
+        if (ins.isEmpty() == false) {
+            // combine equals alongside the existing ors
+            final ZoneId finalZoneId = zoneId;
```

Comment:
This is a leftover from a time when we supported `zoneId`s in other languages. We are not there with ESQL yet, so this could stay as is.